### PR TITLE
Add superfluid unbond partial amount

### PR DESCRIPTION
**Pull request by t4sk**
## What is the purpose of the change

Implement partial unbonding from superfluid staking

Here is an overview of how `SuperfluidUndelegateAndUnbondLock` works.
1. Superfluid undelegate all
2. Superfluid unbond partial amount -> BeginForceUnlock -> Get new "unlocking lock" id
3. Delete synthetic unlocking lock - undo synthetic lock created in step 1
4. Create synthetic unlocking lock for underlying lock from step 2
5. Superfluid re-delegate remaining amount

If the amount to unlock is equal to the underlying locked amount, this function will behave the same as `SuperfluidUndelegate` followed by `SuperfluidUnbondLock`. Otherwise this function will create a new synthetic unlocking lock from the newly created underlying lock.

Related links
https://github.com/osmosis-labs/osmosis/pull/4059#event-8331146656
https://github.com/osmosis-labs/osmosis/issues/3830

## Brief Changelog

- Implements `SuperfluidUndelegateAndUnbondLock`
 

## Testing and Verifying

This change added tests and can be verified as follows:

*(example:)*
 - Added unit test in`x/superfluid/keeper/stake_test.go` 

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes)
  - How is the feature or change documented? (documented)

### CODE REVIEW

**Comment by p0mvn** [comment]
Thanks for the PR. Can you please add a changelog entry for this?

File: x/superfluid/keeper/stake_test.go
Comment (mattverse) [inline x/superfluid/keeper/stake_test.go:469]: isn't lock length always = 1 ? If so, can we put it as a variable and then remove the for loop below/

File: x/superfluid/keeper/stake_test.go
Comment (mattverse) [inline x/superfluid/keeper/stake_test.go:512]: Can we also add check for unstaking synth denom for `tc.splitLockId` and check for staking synth denom in the `!tc.splitLockId` cases?

File: x/superfluid/keeper/stake_test.go
Comment (mattverse) [inline x/superfluid/keeper/stake_test.go:455]: Can we also add edge case testing for this method such as trying to UndelegateAndUnbondLock an existing unlocking lock || trying to UndelegateAndUnbondLock on a lock that still has bonding but is superfluid undelegating and so on?

**Review by mattverse — COMMENTED** [review]
Thanks for the PR! Looks nice 🌮  Left some comments and questions I had + requested some more test cases and checks.

File: x/superfluid/types/codec.go
Comment (t4sk) [reply x/superfluid/types/codec.go:17]: yup any suggestion from the core team?
How about `MsgSuperfluidUnboundLockPartial`?

File: x/superfluid/keeper/stake.go
Comment (t4sk) [reply x/superfluid/keeper/stake.go:332]: I think lockID is correct, but there might be something I misunderstood about the code.

The underlying lock will be split only if the unlock amount is less than the locked amount.
The newly split lock will be an unlocking lock. Original lock will still be locked.

So inside superfluid we need to update in accordance with what happened in the lockup module.
If `lockID = newLockID`, then no lock was split, so the code exits without re-delegating

https://github.com/stakewithus/osmosis/blob/9fcd3c123967f277c90c7ad2f66809568092b5af/x/superfluid/keeper/stake.go#L317

Otherwise `lockID != newLockID`, so a new unlocking lock was created.
`lockID` contains the amount still locked in the `lockup` module.
`newLockID` contains the amount unlocked.

So we re-delegate `lockID`, updating the amount to superfluid stake, and create a synthetic unlocking lock for `newLockID`

File: x/superfluid/keeper/stake.go
Comment (t4sk) [reply x/superfluid/keeper/stake.go:331]: it won't re-delegate if `lockID == newLockID`, it does an early exist here
https://github.com/stakewithus/osmosis/blob/9fcd3c123967f277c90c7ad2f66809568092b5af/x/superfluid/keeper/stake.go#L317

File: x/superfluid/keeper/stake_test.go
Comment (t4sk) [reply x/superfluid/keeper/stake_test.go:469]: Yes I can. Alternatively would you like me to test with multiple locks?

File: x/superfluid/types/codec.go
Comment (mattverse) [reply x/superfluid/types/codec.go:17]: I think `MsgSuperfluidUnboundLockPartial` can cause confusion with `MsgSuperfluidUnbondLock` since its doing much more than `MsgSuperfluidUnbondLock`. I suggest using shorter terms for superfluid, 
`osmosis/superfluid-undelegate-and-unbond-lock` -> `osmosis/sf-undelegate-and-unbond-lock`

File: x/superfluid/keeper/stake.go
Comment (mattverse) [reply x/superfluid/keeper/stake.go:332]: I see I see, this makes sense! Would it be possible to add some of these explanations as inline comments?

**Comment by mattverse** [comment]
@t4sk Would it be possible to re-assign me as reviewer once you think the PR is ready for a second round of review?

**Comment by t4sk** [comment]
@mattverse It's ready for another round of review. But I don't have write access to assign reviewers

File: x/superfluid/keeper/stake_test.go
Comment (mattverse) [inline x/superfluid/keeper/stake_test.go:556]: Can we also add check for total supply of stake denom to make sure that we have burnt the correct amount?

**Review by mattverse — COMMENTED** [review]
Once final test case requested has been added, lgtm

File: proto/osmosis/superfluid/tx.proto
Comment (stackman27) [inline proto/osmosis/superfluid/tx.proto:68]: i dont think castrepeated is necessary when it's just a single coin, i've always seen people use cast repeated with "repeated" key term

```suggestion
   cosmos.base.v1beta1.Coin token_out = 3 [
    (gogoproto.moretags) = "yaml:\"coin\"",
    (gogoproto.nullable) = false
  ];
```

File: x/superfluid/keeper/internal/events/emit.go
Comment (stackman27) [inline x/superfluid/keeper/internal/events/emit.go:130]: Out of curiosity: wouldn't you want to emit the amount of tokens undelegated too?

File: x/superfluid/keeper/stake.go
Comment (stackman27) [inline x/superfluid/keeper/stake.go:317]:
```suggestion
	// Otherwise unbond amount < locked amount, and the underlying lock was split.
```

File: x/superfluid/keeper/stake.go
Comment (stackman27) [inline x/superfluid/keeper/stake.go:341]: I'm a little confused, aren't we re-delegating the newLockID?

File: x/superfluid/keeper/stake.go
Comment (stackman27) [inline x/superfluid/keeper/stake.go:347]: Out of curiosity: Does it matter if we createSythLockup before or after we delegate, I'm following your overview in PR description, and we delegate after creating synthetic lockup

File: x/superfluid/keeper/stake.go
Comment (stackman27) [inline x/superfluid/keeper/stake.go:273]: nit: just wondering if there's a specific reason why you didn't do this here instead a declaring a whole new function? 

```suggestion
		lock, err := k.lk.GetLockByID(ctx, underlyingLockId)
	if err != nil {
		return 0, err
	}
	err = k.validateLockForSF(ctx, lock, sender)
	if err != nil {
		return 0, err
	}
	synthLocks := k.lk.GetAllSyntheticLockupsByLockup(ctx, underlyingLockId)
	if len(synthLocks) != 1 {
		return 0, types.ErrNotSuperfluidUsedLockup
	}
	if !synthLocks[0].IsUnlocking() {
		return 0, types.ErrBondingLockupNotSupported
	}
	return k.lk.BeginForceUnlock(ctx, underlyingLockId, coins)
}
```

**Review by stackman27 — COMMENTED** [review]
Thank you @t4sk for the PR, i left few comments! 🐙

File: x/superfluid/keeper/stake.go
Comment (t4sk) [reply x/superfluid/keeper/stake.go:341]: newLockID contains an underlying unlocking lock. More explained here
https://github.com/osmosis-labs/osmosis/pull/4107#discussion_r1088637404

File: x/superfluid/keeper/stake.go
Comment (t4sk) [reply x/superfluid/keeper/stake.go:347]: I don't think order matters. I can experiment by switching the order of operations

File: x/superfluid/keeper/internal/events/emit.go
Comment (t4sk) [reply x/superfluid/keeper/internal/events/emit.go:130]: I think the event is already emitted by the lock up module
https://github.com/osmosis-labs/osmosis/blob/d28c7412c8d16ed48a20f7b770185d238d83f12b/x/lockup/keeper/lock.go#L247-L249

File: x/superfluid/keeper/stake.go
Comment (t4sk) [reply x/superfluid/keeper/stake.go:273]: It's because the function `unbondLock` is also used in `SuperfluidUndelegateAndUnbondLock`
https://github.com/stakewithus/osmosis/blob/3930891ded8c4f7ab10b0f854a9368bcdaeaa997/x/superfluid/keeper/stake.go#L308

File: x/superfluid/keeper/stake_test.go
Comment (t4sk) [reply x/superfluid/keeper/stake_test.go:566]: I've added code to check the total supply of OSMO. Please let me know if you were thinking of something else

**Review by mattverse — APPROVED** [review]
Awesome! LGTM 🌮

**Review by stackman27 — COMMENTED** [review]
LGTM! 💯

**Review by stackman27 — APPROVED** [review]

**Comment by p0mvn** [comment]
Merging this since I can see 2 approvals. Thanks everyone!

**Pull request merged**
